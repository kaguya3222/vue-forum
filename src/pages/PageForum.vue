<template>
  <div v-if="asyncDataStatus_ready" class="forum-wrapper">
    <div class="col-full push-top">
      <div class="forum-header">
        <div class="forum-details">
          <h1>{{ forum.name }}</h1>
          <p class="text-lead">{{ forum.description }}</p>
        </div>
        <router-link
          :to="{
            name: 'ThreadCreate',
            params: { forumId: this.forum['.key'] }
          }"
          class="btn-green btn-small"
          >Start a thread</router-link
        >
      </div>
    </div>

    <div class="col-full push-top">
      <thread-list :threads="forumThreads" />
    </div>
  </div>
</template>

<script>
import ThreadList from "../components/ThreadList";
import { mapGetters, mapActions } from "vuex";
import asyncDataStatus from "../mixins/asyncDataStatus";

export default {
  name: "PageForum",
  components: {
    "thread-list": ThreadList
  },
  mixins: [asyncDataStatus],
  props: {
    id: {
      required: true,
      type: String
    }
  },
  computed: {
    ...mapGetters({
      threads: "threads/items",
      forums: "forums/items"
    }),
    forum() {
      return this.forums[this.id];
    },
    forumThreads() {
      return Object.values(this.threads).filter(
        thread => thread.forumId === this.id
      );
    }
  },
  methods: {
    ...mapActions("forums", ["fetchForum"]),
    ...mapActions("threads", ["fetchThreads"]),
    ...mapActions("users", ["fetchUser"])
  },
  created() {
    this.fetchForum({ id: this.id })
      .then(forum => {
        return this.fetchThreads({ ids: forum.threads });
      })
      .then(threads => {
        return Promise.all(
          threads.map(thread => this.fetchUser({ id: thread.userId }))
        );
      })
      .then(() => {
        this.asyncDataStatus_fetched();
      });
  }
};
</script>

<style scoped>
.forum-wrapper {
  width: 100%;
}
</style>
