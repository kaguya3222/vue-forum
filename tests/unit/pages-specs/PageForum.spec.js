import { createLocalVue, shallowMount } from "@vue/test-utils";
import PageForum from "../../../src/pages/PageForum";
import ThreadList from "../../../src/components/ThreadList";
import mockedSourceData from "../mocks/mockedSourceData";
import Vuex from "vuex";
import sourceStore from "../../../src/store";
import threadsGetters from "@/store/modules/threads/getters";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("PageForum", () => {
  const store = new Vuex.Store({
    state: sourceStore.state,
    getters: {
      ...threadsGetters
    }
  });
  const wrapper = shallowMount(PageForum, {
    propsData: {
      id: Object.values(mockedSourceData.forums)[1][".key"]
    },
    stubs: ["router-link"],
    localVue,
    store
  });
  test("Shows forum data on forum page", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Accepts ThreadList as a child component", () => {
    expect(wrapper.contains(ThreadList)).toBe(true);
  });
  test("Takes forum object from source data with given id", () => {
    expect(wrapper.vm.forum[".key"]).toBe(wrapper.vm.id);
  });
  test("Creates threads array of certain forum", () => {
    const thread = Object.values(mockedSourceData.threads)[1];
    const hasTread = wrapper.vm.forumThreads.some(
      forumThread => forumThread[".key"] === thread[".key"]
    );
    expect(hasTread).toBe(true);
  });
});
