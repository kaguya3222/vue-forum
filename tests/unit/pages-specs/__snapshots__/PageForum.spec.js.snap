// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageForum Shows forum data on forum page 1`] = `
<div class="forum-wrapper">
  <div class="col-full push-top">
    <div class="forum-header">
      <div class="forum-details">
        <h1>Questions &amp; Feedback</h1>
        <p class="text-lead">Ask your questions and drop your feedback here</p>
      </div>
      <router-link-stub to="[object Object]" class="btn-green btn-small">Start a thread</router-link-stub>
    </div>
  </div>
  <div class="col-full push-top">
    <thread-list-stub threads="[object Object]"></thread-list-stub>
  </div>
</div>
`;
